kernel/frame_allocator/src/lib.rs

```diff
@@ -262,113 +321,330 @@ pub enum MemoryRegionType {
+///
+/// This exists to break the cyclic dependency chain between this crate and
+/// the `page_table_entry` crate, since `page_table_entry` must depend on types
+/// from this crate in order to enforce safety when modifying page table entries.
+pub(crate) fn into_unmapped_frames(frames: FrameRange) -> UnmappedFrames {
+    let typ = if contains_any(&RESERVED_REGIONS.lock(), &frames) {
+        MemoryRegionType::Reserved
+    } else {
+        MemoryRegionType::Free
+    };
+    Frames{ typ, frames }
+}
+
+
+impl<const S: MemoryState> Drop for Frames<S> {
+    fn drop(&mut self) {
+        match S {
+            MemoryState::Free => {
+                if self.size_in_frames() == 0 { return; }
+        
+                let mut free_frames: FreeFrames = Frames {
+                    typ: self.typ,
+                    frames: self.frames.clone(),
+                };
+        
+                let mut list = if free_frames.typ == MemoryRegionType::Reserved {
+                    FREE_RESERVED_FRAMES_LIST.lock()
+                } else {
+                    FREE_GENERAL_FRAMES_LIST.lock()
+                };        
+            
+                match &mut list.0 {
+                    // For early allocations, just add the deallocated chunk to the free pages list.
+                    Inner::Array(_) => {
+                        if list.insert(free_frames).is_ok() {
+                            return;
+                        }
+                    }
+                    
+                    // For full-fledged deallocations, use the entry API to efficiently determine if
+                    // we can merge the deallocated frames with an existing contiguously-adjacent chunk
+                    // or if we need to insert a new chunk.
+                    Inner::RBTree(ref mut tree) => {
+                        let mut cursor_mut = tree.lower_bound_mut(Bound::Included(free_frames.start()));
+                        if let Some(next_frames_ref) = cursor_mut.get() {
+                            if *free_frames.end() + 1 == *next_frames_ref.start() {
+                                // extract the next chunk from the list
+                                let mut next_frames = cursor_mut
+                                    .replace_with(Wrapper::new_link(Frames::empty()))
+                                    .expect("BUG: couldn't remove next frames from free list in drop handler")
+                                    .into_inner();
+
+                                // trace!("Prepending {:?} onto beg of next {:?}", free_frames, next_frames);
+                                if next_frames.merge(free_frames).is_ok() {
+                                    // trace!("newly merged next chunk: {:?}", next_frames);
+                                    // now return newly merged chunk into list
+                                    match cursor_mut.replace_with(Wrapper::new_link(next_frames)) { 
```

Comment:
this whole sequence is quite inefficient, in two main ways:

1. We replace the frames at the cursor with an empty frames. There's no real point in doing this; we might as well just remove the frames from the list, modify (extend) them, and then re-insert them at the current cursor. That's as efficient as we can get.
2. Using `merge` is redundant -- we have already checked the merge condition here in L497, so it's a redundant check. The corresponding panic in L513 is also redundant and unnecessary, since we can directly modify the inner range of the frames object that we removed from the list/cursor above.

To summarize, we should just do the following:
1. Remove the frames in L499.
2. Directly modify those frames instead of redundantly calling `merge` on those frames.
3. Re-insert the modified frames object back into the rbtree using the current cursor.

Please also apply this change to the code block starting on L517.

Reply:
Point 1 makes sense, thanks for pointing that out.
For point 2, I can replace it for now, but once a `Frames` object is composed of a `TrustedChunk`, then we will have to use the merge function since we cannot directly set the range of a `TrustedChunk`.

Reply:
Okay, if that's the case for #2, then don't bother with it until we are reviewing the `TrustedChunk` PR.

However, I'd question why that is _required_, or if it's just required by the current `TrustedChunk` design. Regardless, it should be possible to refactor that code sequence to avoid the redundant conditional (in L497 and in `merge()`), which then will subsequently avoid the unreachable error condition in L513.

Reply:
I've replaced `replace_with` with `remove`. For now I'll keep it at that.

But for future discussion on the `TrustedChunk` design:
Checking the merge conditions and performing the merge needs to take place in the same function, as that's the granularity of verification. So for that, we need mutable access to the contiguous chunk in the rbtree which is only possible by removing it from the tree. So the sequence could be to `remove frames from the tree -> merge -> reinsert into tree `rather than `check merge conditions -> remove from tree -> merge -> reinsert`, but that would lead to the overhead of removing an object from the rbtree in every drop handler.

Reply:
Okay cool.

There's an easy way to get around that, just have another method for `TrustedChunk` that acts as proof-of-work, sort of like a `VerifiedMergeable` type that is returned from that method that just performs the check. Then, when you pass that into another `merge_verified()`/pre-checked function, you know you don't need to re-evaluate the conditional. But we can get to that later.